lib/resty/rediscluster.lua

```diff
@@ -347,35 +376,66 @@ function _M.new(_, config)
 end
 
 
-local function pick_node(self, serv_list, slot, magic_radom_seed)
+local function pick_node(self, serv_list, magic_radom_seed, select_ip)
     local host
     local port
     local slave
     local index
+
     if #serv_list < 1 then
-        return nil, nil, nil, "serv_list for slot " .. slot .. " is empty"
+        return nil, nil, nil, "serv_list is empty"
     end
+
     if self.config.enable_slave_read then
         if magic_radom_seed then
             index = magic_radom_seed % #serv_list + 1
         else
             index = math.random(#serv_list)
         end
+
+        -- compatible check if Redis 7 cluster created with IP directly
+        if not serv_list[index].ip then
+            index = index - 1
+        end
+
         host = serv_list[index].ip
         port = serv_list[index].port
-        --cluster slots will always put the master node as first
-        if index > 1 then
-            slave = true
+
+        -- cluster slots will always put the master node before slave nodes
+        if redis_version >= REDIS_VERSION_70 then
+            slave = index > CONSTANT_TWO
+
         else
-            slave = false
+            slave = index > 1
         end
-        --ngx.log(ngx.NOTICE, "pickup node: ", c(serv_list[index]))
+
     else
-        host = serv_list[1].ip
-        port = serv_list[1].port
+        index = 1
+        if redis_version >= REDIS_VERSION_70 then
+            if select_ip and type(select_ip) ~= "number" then
+              return nil, nil, nil, "select_ip must be a number"
+            end
+            if not select_ip or select_ip < 1 then
+                index = CONSTANT_TWO
+
+            elseif select_ip > CONSTANT_TWO then
+                index = select_ip % CONSTANT_TWO + 1
```

Comment:
Please remove the CONSTANT_TWO constant as it only confuses matters.  Explain in a comment why in node selection works like this in Redis 7.

Reply:
Done.